### PR TITLE
Sync framework

### PR DESCRIPTION
**Pull request by adam-stokes**
This is a refactor to support syncing all snaps, the three main components of this are:

- Pulls in all latest tags from upstream and makes sure our snap git repos are synced with the proper tag and snapcraft yaml substitutions
- Syncs all latest branches with their latest releases in the snap store and making sure tracks are updated. 
- Makes sure the latest stable release is always synced into the latest tracks
- Makes sure the next development release is always synced in the next major.minor/edge track

Basically everything is handled automatically, no more needing to manually kick off jobs to ensure all latest supported releases are where they need to be.

### CODE REVIEW

File: cilib/models/repos/__init__.py [hunk 1]
```diff
@@ -48,17 +47,21 @@ def branches(self, **subprocess_kwargs):
         """Grabs remote branches"""
         return git.remote_branches(self.repo, **subprocess_kwargs)
 
+    def latest_branch_from_major_minor(self, major_minor, exclude_pre=False):
+        """Grabs latest known branch semver for a major.minor release"""
+        return self._latest_from_semver(self.branches, major_minor, exclude_pre)
+
+    def latest_tag_from_major_minor(self, major_minor, exclude_pre=False):
+        """Grabs latest known tag semver for a major.minor release"""
+        return self._latest_from_semver(self.tags, major_minor, include_prerelease)
```
Comment (johnsca) [inline cilib/models/repos/__init__.py:56]: Should that be `exclude_pre`?

File: cilib/models/repos/__init__.py [hunk 2]
```diff
@@ -70,3 +73,31 @@ def tags_subset_semver_point(self, alt_model, starting_semver):
...
+        for _semver in semvers:
+            try:
+                semver_version = version.parse(_semver)
+                if exclude_pre and semver_version.prerelease is not None:
+                    continue
+                if major_minor == f"{semver_version.major}.{semver_version.minor}":
+                    _semvers.append(str(semver_version))
+            except:
+                continue
+        return str(max(map(semver.VersionInfo.parse, _semvers)))
```
Comment (johnsca) [inline cilib/models/repos/__init__.py:91]: Could this just be `version.parse` and drop the semver import on this file?

File: jobs/sync-upstream/sync.py [hunk 3]
```diff
@@ -249,7 +260,7 @@ def sync_internal_tags(dry_run):
 
     # List of tuples containing upstream, downstream models and a starting semver
     repos_map = [
-        (UpstreamKubernetesRepoModel(), InternalKubernetesRepoModel(), "v1.19.0")
+        (UpstreamKubernetesRepoModel(), InternalKubernetesRepoModel(), "v1.16.0")
```
Comment (johnsca) [inline jobs/sync-upstream/sync.py:263]: Point this to the enum instead of hardcoding it.

**Review by johnsca — APPROVED** [review]
LGTM, though the complexity of all of this still leaves me a little nervous about turning it loose against the snap store.  Maybe a dry-run functionality would help with that, doing everything except actually updating the snap store?

**Comment by adam-stokes** [comment]
Heres some output with latest dryrun support enabled:

```
2020-10-22 14:10:43.932 | INFO     | cilib.log:info:37 - [kube-controller-manager] > Versions match 1.19.3 == 1.19.3, not building a new snap
2020-10-22 14:10:43.932 | INFO     | cilib.log:info:37 - [kube-controller-manager] > Checking snaps in version 1.19 for arch ppc64el
2020-10-22 14:10:43.933 | INFO     | cilib.log:info:37 - [kube-controller-manager] Found snap version 1.19.3 at revision 1677 for 1.19/edge
2020-10-22 14:10:46.568 | INFO     | cilib.log:info:37 - [kube-controller-manager] Latest branch version 1.19.3
2020-10-22 14:10:46.569 | INFO     | cilib.log:info:37 - [kube-controller-manager] > Versions match 1.19.3 == 1.19.3, not building a new snap
2020-10-22 14:10:46.569 | INFO     | cilib.log:info:37 - [kube-controller-manager] > Checking snaps in version 1.19 for arch s390x
2020-10-22 14:10:46.570 | INFO     | cilib.log:info:37 - [kube-controller-manager] Found snap version 1.19.3 at revision 1675 for 1.19/edge
2020-10-22 14:10:49.096 | INFO     | cilib.log:info:37 - [kube-controller-manager] Latest branch version 1.19.3
2020-10-22 14:10:49.096 | INFO     | cilib.log:info:37 - [kube-controller-manager] > Versions match 1.19.3 == 1.19.3, not building a new snap
2020-10-22 14:10:49.097 | INFO     | cilib.log:info:37 - [kube-controller-manager] > Checking snaps in version 1.19 for arch arm64
2020-10-22 14:10:49.098 | INFO     | cilib.log:info:37 - [kube-controller-manager] Found snap version 1.19.2 at revision 1674 for 1.19/edge
2020-10-22 14:10:53.190 | INFO     | cilib.log:info:37 - [kube-controller-manager] Latest branch version 1.19.3
2020-10-22 14:10:53.191 | INFO     | cilib.log:info:37 - [kube-controller-manager] Found new branch 1.19.3 > 1.19.2, building new snap
2020-10-22 14:10:53.191 | INFO     | cilib.log:info:37 - [kube-controller-manager] > Creating recipe for {'name': 'kube-controller-manager', 'owner': 'k8s-jenkaas-admin', 'version': '1.19', 'branch': 'v1.19.3', 'repo': 'git+ssh://**@git.launchpad.net/snap-kube-controller-manager', 'track': ['1.19/stable', '1.19/candidate', '1.19/beta', '1.19/edge']}
```
So we can expand our debugging and have it simulate the calls here pretty easily
